pyro/distributions/multivariate_normal.py

```diff
@@ -0,0 +1,114 @@
+                raise ValueError("The covariance matrix must be a matrix, but got covariance_matrix.size() = {}".format(
+                                    loc.size()))
+            self.sigma = covariance_matrix
+            self.sigma_cholesky = Variable(
+                torch.potrf(covariance_matrix.data)) if torch.__version__ < '0.3.0' else torch.potrf(covariance_matrix)
+        else:
+            if not scale_tril.dim() == 2:
+                raise ValueError("The Cholesky decomposition of the covariance matrix must be a matrix, "
+                                 "but got scale_tril.size() = {}".format(
+                                    loc.size()))
+            self.sigma = torch.mm(scale_tril.transpose(0, 1), scale_tril)
+            self.sigma_cholesky = scale_tril
+        if loc.dim() > 1:
+            raise ValueError("The mean must be a vector, but got loc.size() = {}".format(loc.size()))
+
+        super(MultivariateNormal, self).__init__(*args, **kwargs)
+
+    def batch_shape(self, x=None):
+        mu = self.mu.expand(self.batch_size, *self.mu.size()).squeeze(0)
+        if x is not None:
+            if x.size()[-1] != mu.size()[-1]:
+                raise ValueError("The event size for the data and distribution parameters must match.\n"
+                                 "Expected x.size()[-1] == self.mu.size()[0], but got {} vs {}".format(
+                                    x.size(-1), mu.size(-1)))
+            try:
+                mu = mu.expand_as(x)
+            except RuntimeError as e:
+                raise ValueError("Parameter `mu` with shape {} is not broadcastable to "
+                                 "the data shape {}. \nError: {}".format(mu.size(), x.size(), str(e)))
+
+        return mu.size()[:-1]
+
+    def event_shape(self):
+        return self.mu.size()[-1:]
+
+    def sample(self):
+        """
+        Generate a sample with the specified covariance matrix and mean.
+        Differentiation wrt. to the covariance matrix is only supported on PyTorch version 0.3.0 or higher.
+        Ref: :py:meth:`pyro.distributions.distribution.Distribution.sample`
+
+        """
+        batch_size = self.batch_size
+        uncorrelated_standard_sample = Variable(torch.randn(batch_size, *self.mu.size()).type_as(self.mu.data))
+        transformed_sample = self.mu + torch.mm(uncorrelated_standard_sample, self.sigma_cholesky)
+        return transformed_sample if self.reparameterized else transformed_sample.detach()
+
+    def batch_log_pdf(self, x):
+        batch_size = x.size()[0] if len(x.size()) > len(self.mu.size()) else 1
+        batch_log_pdf_shape = self.batch_shape(x) + (1,)
+        x = x.view(batch_size, *self.mu.size())
+        normalization_factor = torch.log(
+            self.sigma_cholesky.diag()).sum() + (self.mu.size()[0] / 2) * np.log(2 * np.pi) if self.normalized else 0
+        sigma_inverse = torch.inverse(self.sigma) if self.use_inverse_for_batch_log else \
```

Comment:
As I understand, it's generally a good idea to avoid computing an inverse when solving an equation: if you have a cholesky you should use a Cholesky solver, or if you have a matrix you should use a general matrix solver. Fro what you know of PyTorch, is this not yet possible?

Reply:
It looks like you could do this with `torch.gesv()`: http://pytorch.org/docs/master/torch.html#torch.gesv

Reply:
I was thinking that it is easier to use `torch.inverse()` because `torch.gesv()` is not batch solver.

Reply:
Yeah, that is really a problem. It looks like using `torch.gesv()` only brings performance gains if the batch size is considerably smaller than that of the covariance matrix.